### PR TITLE
TS-4072 Diagnostic log rolling races

### PR DESCRIPTION
**Pull request by danobi**
This patch prevents race conditions when a diagnostic log rolls. The patch is a little longer than I had hoped for because of error handling conditions.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Looks good to me. Dan, Jason, and I discussed this at length prior to the work and this looks in line with our concensus.

File: lib/ts/Diags.cc [hunk 1]
```diff
@@ -642,10 +642,10 @@ Diags::should_roll_diagslog()
...
 
   log_log_trace("should_roll_diagslog() was called\n");
-  log_log_trace("rolling_enabled = %d, output_rolling_size = %d, output_rolling_interval = %d\n", diagslog_rolling_enabled,
-                diagslog_rolling_size, diagslog_rolling_interval);
-  log_log_trace("RollingEnabledValues::ROLL_ON_TIME = %d\n", RollingEnabledValues::ROLL_ON_TIME);
-  log_log_trace("time(0) - last_roll_time = %d\n", time(0) - diagslog_time_last_roll);
+  log_log_trace("should_roll_diagslog(): rolling_enabled = %d, output_rolling_size = %d, output_rolling_interval = %d\n",
+                diagslog_rolling_enabled, diagslog_rolling_size, diagslog_rolling_interval);
+  log_log_trace("should_roll_diagslog(): RollingEnabledValues::ROLL_ON_TIME = %d\n", RollingEnabledValues::ROLL_ON_TIME);
+  log_log_trace("should_roll_diagslog(): time(0) - last_roll_time = %d\n", time(0) - diagslog_time_last_roll);
```
Comment (jpeach) [inline lib/ts/Diags.cc]: Use __func__ to generate the function name.

File: lib/ts/Diags.cc [hunk 2]
```diff
@@ -853,29 +867,39 @@ Diags::set_stderr_output(const char *_bind_stderr)
...
-    delete stderr_log;
+    fprintf(stderr, "[Warning]: stderr is currently not bound to anything\n");
+    delete new_stderr_log;
+    lock();
     stderr_log = NULL;
+    unlock();
     return false;
   }
-  if (!stderr_log->m_fp) {
+  if (!new_stderr_log->m_fp) {
```
Comment (jpeach) [inline lib/ts/Diags.cc]: ``new_stderr_log->is_open()``

File: lib/ts/Diags.cc [hunk 3]
```diff
@@ -816,29 +820,39 @@ Diags::set_stdout_output(const char *_bind_stdout)
...
     stdout_log = NULL;
+    unlock();
     return false;
   }
 
-  return rebind_stdout(fileno(stdout_log->m_fp));
+  // now exchange the stdout_log pointer
+  lock();
+  stdout_log = new_stdout_log;
+  bool ret = rebind_stdout(fileno(new_stdout_log->m_fp));
```
Comment (jpeach) [inline lib/ts/Diags.cc]: If this fails, are we in a well-defined state? No caller ever checks the return value of this ... what would happen if it fails?

File: lib/ts/Diags.cc [hunk 4]
```diff
@@ -816,29 +820,39 @@ Diags::set_stdout_output(const char *_bind_stdout)
...
 
-  if (stdout_log) {
-    delete stdout_log;
-    stdout_log = NULL;
-  }
-
-  // create backing BaseLogFile for stdout
-  stdout_log = new BaseLogFile(_bind_stdout);
+  BaseLogFile *old_stdout_log = stdout_log;
+  BaseLogFile *new_stdout_log = new BaseLogFile(_bind_stdout);
```
Comment (jpeach) [inline lib/ts/Diags.cc]: Can we find a better argument name than ``_bind_stdout``? Maybe ``path`` or ``stdout_path``?

File: lib/ts/Diags.cc [hunk 5]
```diff
@@ -716,14 +722,12 @@ Diags::should_roll_outputlog()
...
 
-  /*
   log_log_trace("should_roll_outputlog() was called\n");
-  log_log_trace("rolling_enabled = %d, output_rolling_size = %d, output_rolling_interval = %d\n", outputlog_rolling_enabled,
-                outputlog_rolling_size, outputlog_rolling_interval);
-  log_log_trace("RollingEnabledValues::ROLL_ON_TIME = %d\n", RollingEnabledValues::ROLL_ON_TIME);
-  log_log_trace("time(0) - last_roll_time = %d\n", time(0) - outputlog_time_last_roll);
-  log_log_trace("stdout_log = %p\n", stdout_log);
-  */
+  log_log_trace("should_roll_outputlog(): rolling_enabled = %d, output_rolling_size = %d, output_rolling_interval = %d\n",
```
Comment (jpeach) [inline lib/ts/Diags.cc]: I think ``log_log_trace`` ought to be something you can set at runtime. Probably an environment variable or something that sets a global flag?

File: lib/ts/Diags.cc [hunk 6]
```diff
@@ -816,29 +820,39 @@ Diags::set_stdout_output(const char *_bind_stdout)
...
-  stdout_log = new BaseLogFile(_bind_stdout);
+  BaseLogFile *old_stdout_log = stdout_log;
+  BaseLogFile *new_stdout_log = new BaseLogFile(_bind_stdout);
 
   // on any errors we quit
-  if (!stdout_log || stdout_log->open_file() != BaseLogFile::LOG_FILE_NO_ERROR) {
+  if (!new_stdout_log || new_stdout_log->open_file() != BaseLogFile::LOG_FILE_NO_ERROR) {
     fprintf(stderr, "[Warning]: unable to open file=%s to bind stdout to\n", _bind_stdout);
-    delete stdout_log;
+    fprintf(stderr, "[Warning]: stdout is currently not bound to anything\n");
```
Comment (jpeach) [inline lib/ts/Diags.cc]: This would be better as ``log_log_error()``.

**Comment by jpeach** [comment]
OK the locking makes sense to me I think. You have to hold the lock because you want to serialize WRT printing output.

``Diags::rebind_stderr`` and ``Diags::rebind_stdout`` don't need to be part of the class interface. You can replace both of these with a single local static helper:
```C
bool rebind(int oldfd, int newfd) {
    /* ... error handling ... */
    return dup2(newfd, oldfd) == 0;
}
```

Please write unit tests for this.

**Comment by SolidWallOfCode** [comment]
Yes. We looked at various approaches and decided on this one because there was already a lock around access to the file descriptor. Locking the update to that value should prevent the race condition and potential use of a deleted file handle.

File: lib/ts/Diags.cc (hunk 5 above)
Comment (SolidWallOfCode) [inline lib/ts/Diags.cc]: That's outside the scope of this fix. It should be in a separate patch on a different ticket.

Comment (danobi) [inline lib/ts/Diags.cc]: Yeah I was thinking about that when I first wrote the function. I'm not sure what the best way to do it would be. Right now it's compile time defined with a `#define`.

**Comment by danobi** [comment]
@jpeach Is there a preferred framework for unit testing? (ie how/where do I put the tests?)

**Comment by SolidWallOfCode** [comment]
While I think unit tests are in general a good idea, I'm not sure how that could be done in this case and not in the time frame for committing this to 6.2. I expect tests will need to wait for a better test framework (via TSQA or successor) because of the interaction with the file system. This is especially acute due to the lack of any pre-existing unit tests which could be extended.

File: lib/ts/Diags.cc (hunk 3 above)
Comment (danobi) [inline lib/ts/Diags.cc]: It should be fine.

At that point in code, `m_fp` is guaranteed to be pointing to some open file. If `rebind_stdout()` fails, it's because (1) `fileno()` failed or (2) `dup2()` failed. If either of those happen, then there's really nothing we can do about it (AFAICT). So ATS will just plod along without `stdout` or `stderr` being bound to anything. Fortunately nobody should be writing directly to either of those streams so we should be in reasonable shape even if `rebind_...()` fails. Hopefully the next time the log files roll the error will correct itself.

File: lib/ts/Diags.cc (hunk 5 above)
Comment (jpeach) [inline lib/ts/Diags.cc]: Yes; the compilation time option means that is it subject to bitrot and that few people will be able ti use it. I'm OK with addressing this as a separate change, but I think it is worth addressing.

File: lib/ts/Diags.cc (hunk 3 above)
Comment (jpeach) [inline lib/ts/Diags.cc]: I think there's a couple of issues here. First, if the error can't (or shouldn't) reasonably be handled by the caller, then there's little value in trying to propagate it. 

``fileno()`` can't fail unless there's a programmer error. and `dup2()`` can't fail unless there is a programmer error or we run out of file descriptors. If you look at ``rebind_stderr()``, we don't even propagate the return value of ``dup2()``.

I think that this should be handled by a ``ink_release_assert``.

File: lib/ts/Diags.cc (hunk 5 above)
Comment (SolidWallOfCode) [inline lib/ts/Diags.cc]: +1

**Comment by jpeach** [comment]
I don't see any reason to block writing tests on some proposed framework. All you really need to write a test is a hook to run it (which we have) and a way to check state (which we have). I think that is is possible to write an actual unit test for this, focusing on just the ``Diags`` class, but for now I'd also be ok with a shell-script or something like what we used for testing access log rolling in the first iteration of TSQA.

Off the top of my head, how about a test that starts a few threads writing diagnostics, then a second set of threads that continually rebinds ``stderr`` and ``stdout`` to ``/dev/null``?

**Comment by danobi** [comment]
@jpeach Thanks for the comments.

I've addressed the relevant issues you brought up in the force pushed commits. I've also gone ahead and made a separate ticket with the rest of the issues [Jira ticket](https://issues.apache.org/jira/browse/TS-4354).

**Comment by atsci** [comment]
Can one of the admins verify this patch?

**Comment by zwoop** [comment]
[approve ci]

**Comment by atsci** [comment]
Linux (CentOS7) build finished successfully. Details on https://ci.trafficserver.apache.org/job/Github-Linux/17/

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by zwoop** [comment]
One more round of builds [approve ci].

**Comment by atsci** [comment]
Linux (CentOS7) build failed! Details on https://ci.trafficserver.apache.org/job/Github-Linux/34/

**Comment by atsci** [comment]
FreeBSD v10 build finished successfully. Details on https://ci.trafficserver.apache.org/job/Github-FreeBSD/17/

**Comment by zwoop** [comment]
Upgraded git on all the build boxes, so try again [approve ci].

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by zwoop** [comment]
This needs a rebase and new clang-format indentation. It fails with

https://ci.trafficserver.apache.org/view/github/job/Github-Linux/67/

**Comment by danobi** [comment]
@zwoop I've gone ahead and done that. I don't know how (or if I can) to run the CI builds so I'll just leave that to you.

**Comment by zwoop** [comment]
Can you squash the commits together too? We don't want intermediary commits that are failing (they make git bisect difficult to deal with).

**Comment by danobi** [comment]
@zwoop done

File: lib/ts/Diags.cc [hunk 7]
```diff
@@ -625,11 +625,14 @@ Diags::should_roll_diagslog()
         fflush(diags_log->m_fp);
         if (diags_log->roll()) {
```
Comment (chaiman) [inline lib/ts/Diags.cc]: The old file  don't need to  be closed in  diags_log->roll(). close old file should at destructor of old_diags.  in addition,
this will cause crash because of writing to closed file

File: lib/ts/Diags.cc [hunk 8]
```diff
@@ -642,11 +645,14 @@ Diags::should_roll_diagslog()
...
```
Comment (chaiman) [inline lib/ts/Diags.cc]: same problem

**Comment by danobi** [comment]
@chaiman I see what you mean. I'll fix that in a separate ticket.

File: lib/ts/Diags.cc (hunk 7 above)
Comment (oknet) [inline lib/ts/Diags.cc]: agree with u,  old file will be closed by ~~LINE 635: delete old_diags;~~.

**Comment by zwoop** [comment]
.

**Comment by danobi** [comment]
Fixed

**Comment by PSUdaemon** [comment]
Are there still outstanding issues here or is it possible to merge this now?

**Comment by danobi** [comment]
@PSUdaemon There's not any issues I'm aware of at the moment.

**Comment by zwoop** [comment]
Well, this slipped through the net. @danobi  Can you please make an update to the PR, that does not have merge conflicts?

**Comment by danobi** [comment]
@zwoop clang-formatted & rebased

**Comment by zwoop** [comment]
[approve ci] @PSUdaemon @jpeach thoughts?

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/474/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/371/ for details.

**Comment by zwoop** [comment]
Time to land this?
